File: boing.py
import numpy as np
import math
from pathlib import Path

sLat = 180/8
sLong = 180/8
radius = 220.0
rotAngle = -20
aspectConst = 1.15


def cos(deg):
    return math.cos(math.radians(deg))

def sin(deg):
    return math.sin(math.radians(deg))

cosra = cos(rotAngle)
sinra = sin(rotAngle)

data_folder = Path("BoingSrc/source/")

def drawGrid():

    latfile = data_folder / "grid.h"
    
    with open(latfile, 'w') as f:

        latSeg = 0

        #   VIA_t1_cnt_lo = MOVE_SCALE;
        ##   Moveto_d(90, -108);
        stepy = -18
#        for x in range(0, 6, 1):
            #print(, file=f )   # print out move command of stepy
            #print(, file=f )   # print draw command of 0, 127
            #print(, file=f )   # print out move command of stepy


# // Grid
# if (1)
# {
#   Intensity_a(0x34);
#   Reset0Ref();
#   VIA_t1_cnt_lo = MOVE_SCALE;
#   Moveto_d(90, -108);
#   VIA_t1_cnt_lo = 248;  // scale

# #define STEPY -18

#       Draw_Line_d(0,127);

#       Moveto_d(STEPY, 0);
#       Draw_Line_d(0,-127);

#       Moveto_d(STEPY, 0);
#       Draw_Line_d(0,127);

#       Moveto_d(STEPY, 0);
#       Draw_Line_d(0,-127);

#       Moveto_d(STEPY, 0);
#       Draw_Line_d(0,127);

#       Moveto_d(STEPY, 0);
#       Draw_Line_d(0,-127);

#       Moveto_d(STEPY, 0);
#       Draw_Line_d(0,127);


# #define STEPX  -18
#       Reset0Ref();
#       Moveto_d(-55, 64);
#       VIA_t1_cnt_lo = 248;  // scale
#       Draw_Line_d(108,0);
#       Moveto_d(0, STEPX);
#       Draw_Line_d(-108,0);
#       Moveto_d(0, STEPX);
#       Draw_Line_d(108,0);

#       Moveto_d(0, STEPX);
#       Draw_Line_d(-108,0);
#       Moveto_d(0, STEPX);
#       Draw_Line_d(108,0);

#       Moveto_d(0, STEPX);
#       Draw_Line_d(-108,0);
#       Moveto_d(0, STEPX);
#       Draw_Line_d(108,0);

#       Moveto_d(0, STEPX);
#       Draw_Line_d(-108,0);
#     }



def drawCircumference():

    circfile = data_folder / "circ.h"
    
    with open(circfile, 'w') as f:

        print ("const signed char ball[] = {", file=f)
        print ("15,", file=f)

        x1 = 0
        y1 = radius

        rx1 = x1 * cosra - y1 * sinra
        ry1 = x1 * sinra + y1 * cosra

        for deg in np.arange (0.0,360.0,sLong):
            x2 = sin(deg + sLong) * radius
            y2 = cos(deg + sLong) * radius
        
            # rotate
            rx2 = x2 * cosra - y2 * sinra
            ry2 = x2 * sinra + y2 * cosra
            
            dx = (rx2 - rx1)
            dy = (ry2 - ry1)

            print(str(int(dy)) +  ",", str(int(dx * aspectConst)) + ",", file=f )  # Full globe if using 360

            rx1 = rx2
            ry1 = ry2

        print ("};", file=f)

def drawLatitude():

    latfile = data_folder / "lat.h"
    
    with open(latfile, 'w') as f:

        scaledRadius = radius / 3.82

        latSeg = 0

        for longDeg in np.arange(sLong, 180.0, sLong ):

            print ("const signed char latSeg" + str(latSeg) + "[] = {", file=f)
            print ("0,", file=f)        

            x1 = 0
            y1 = scaledRadius

            x2 = sin(longDeg) * scaledRadius
            y2 = cos(longDeg) * scaledRadius

            rx1 = x1 * cosra - y1 * sinra
            ry1 = x1 * sinra + y1 * cosra

            rx2 = x2 * cosra - y2 * sinra
            ry2 = x2 * sinra + y2 * cosra

            dx = (rx2 - rx1)
            dy = (ry2 - ry1)

            print(str(int(dy)) +  ",", str(int(dx * aspectConst)) + ",", file=f )

            doffset = (latSeg+1) * 45
            x1 = sin(longDeg-doffset) * scaledRadius
            y1 = cos(longDeg-doffset) * scaledRadius

            rx1 = x1 * cosra - y1 * sinra
            ry1 = x1 * sinra + y1 * cosra

            #rx2 = x2 * cosra - y2 * sinra
            #ry2 = x2 * sinra + y2 * cosra

            dx = (rx1 - rx2)
            dy = (ry1 - ry2)

            print(str(int(dy)) +  ",", str(int(dx * aspectConst)) + ",", file=f )

            print ("};", file=f)
            latSeg = latSeg + 1

def drawLongitude():

    seg = 0

    longfile = data_folder / "long.h"
    
    with open(longfile, 'w') as f:

        for ani in range(0, 22, 2):

            print ("const signed char long" + str(seg) + "[] = {", file=f)
            print ("55,", file=f)

            negF = 1

            for latDeg in np.arange((ani + (sLat/2)), (180.0 - sLat/2), sLat ):

            
                if (negF > 0):
                    for deg in np.arange(0.0, (180.001-sLong), sLong):
                        x1 = cos(latDeg) * (radius * sin(deg))
                        y1 = cos(deg) * radius
                        
                        x2 = cos(latDeg) * (radius * sin(deg + sLong))
                        y2 = cos(deg+sLat) * radius

                        rx1 = x1 * cosra - y1 * sinra
                        ry1 = x1 * sinra + y1 * cosra

                        rx2 = x2 * cosra - y2 * sinra
                        ry2 = x2 * sinra + y2 * cosra

                        dx = (rx2 - rx1)
                        dy = (ry2 - ry1)

                        print(str(int(dy)) +  ",", str(int(dx * aspectConst)) + ",", file=f)
                else:
                    for deg in np.arange((180 - sLong),-0.001 , -sLong):
                        x1 = cos(latDeg) * (radius * sin(deg))
                        y1 = cos(deg) * radius
                        
                        x2 = cos(latDeg) * (radius * sin(deg + sLong))
                        y2 = cos(deg+sLat) * radius

                        rx1 = x1 * cosra - y1 * sinra
                        ry1 = x1 * sinra + y1 * cosra

                        rx2 = x2 * cosra - y2 * sinra
                        ry2 = x2 * sinra + y2 * cosra

                        dx = (rx1 - rx2)
                        dy = (ry1 - ry2)

                        print(str(int(dy)) +  ",", str(int(dx * aspectConst)) + ",", file=f)

                negF = -negF

            seg = seg + 1
            print ("};", file=f)

drawCircumference()
drawLatitude()
drawLongitude()

File: boing.js
var sDiv = 8;
var sLat = (180/sDiv);
var sLong = (180/sDiv);
var radius = 90;
var bx = 204;
var by = 171.3;
var n = 3.8;
var dn = 0.16;
var xs = 1.5;
var lpf = 0;
var tlen = 0;


var circleTab = [];

var rotAngle = 0;
var cosra = cos(rotAngle);
var sinra = sin(rotAngle);

frameRate(20);

var pic = createGraphics(400, 400, JAVA2D);

// Define some common functions for working with vectors
//var add = function(a, b) {  return (a.x + b.x, a.y + b.y);};
//var sub = function(a, b) {  return (a.x - b.x, a.y - b.y);};
//var dot = function(a, b) {  return a.x * b.x + a.y * b.y;};
//var hypot2 = function(a, b) { return dot(sub(a, b), sub(a, b));};
    
//var v1 = new PVector(40, 20);
//var v2 = new PVector(25, 50);     

//PVector.multiply = function(a, b) {
//	if (b instanceof PVector){
//	    return new PVector(a.x * b.x, a.y * b.y);
//	}
//	else {
//	    return new PVector(a.x * b, a.y * b);
//	}
//};

var perpdot = function(a, b) {
	return -a.y * b.x + a.x * b.y;
};

//v1.mult(v2);

//var dd = v1.dot(v2);

//debug(dd);
//debug(v1.z);
    

var drawLine = function(x1, y1, x2, y2, debug)
{
    var x = Math.trunc(x1);
    var y = Math.trunc(y1);
    x2 = Math.trunc(x2);
    y2 = Math.trunc(y2);

    var dx = abs(x-x2);
    var dy = abs(y-y2);

    var hy = dy;
    var hx = dx;
    var ddy = 0;
    var ddx = 0;
    var ix = 1;
    var iy = 1;
    
    if (x1 > x2){
        ix = -1;
    }

    if (y1 > y2){
        iy = -1;
    }

    if (dy > dx)
    {
        while(hy)
        {
            point(x, y);
            ddy = ddy - dx;
    
            if (ddy < 0){
                x = x + ix;
                ddy = ddy + dy;
            }    
            y = y + iy;
            hy = hy  - 1;
        }
    } else {

        while(hx)
        {
            point(x, y);        
            ddx = ddx - dy;
    
            if (ddx < 0){
                y = y + iy;
                ddx = ddx + dx;
            }    
            x = x + ix;
            hx = hx - 1;
        }        
    }    
};


var length = function(x1, y1, x2, y2) {
    
    var vx = x1 - x2;
    var vy = y1 - y2;
    
    return sqrt(sq(vx) + sq(vy));
};

var cline = function(x1, y1, x2, y2) {
    line (x1,y1,x2,y2);
    lpf++;
    tlen += length(x1,y1,x2,y2);
};


var lineCircleCollision = function(circlePos, radius, line1, line2) {
    
    var c = new PVector(circlePos.x,circlePos.y);
    var l2 = new PVector(line2.x, line2.y);
    
    c.sub(line1);
    l2.sub(line1);
    
    var t = l2.dot(c) / l2.dot(l2);

    if (t < 0) {
        t = 0;
    }
    else if (t > 1) {
        t = 1;
    }

    l2.mult(t);
    l2.sub(c);

    return l2.dot(l2) <= sq(radius);
};




var fillCircleTable = function(x1, y1, x2, y2)
{
    var x = Math.trunc(x1);
    var y = Math.trunc(y1);
    x2 = Math.trunc(x2);
    y2 = Math.trunc(y2);

    var dx = abs(x-x2);
    var dy = abs(y-y2);

    var hy = dy;
    var hx = dx;
    var ddy = 0;
    var ddx = 0;
    var ix = 1;
    var iy = 1;
    
    if (x1 > x2){
        ix = -1;
    }

    if (y1 > y2){
        iy = -1;
    }

    if (dy > dx)
    {
        while(hy)
        {
            point(x, y);
            circleTab[y] = x;
            ddy = ddy - dx;
    
            if (ddy < 0){
                x = x + ix;
                ddy = ddy + dy;
            }    
            y = y + iy;
            hy = hy  - 1;
        }
    } else {

        while(hx)
        {
            point(x, y);
            circleTab[y] = x;
            
            ddx = ddx - dy;
    
            if (ddx < 0){
                y = y + iy;
                ddx = ddx + dx;
            }    
            x = x + ix;
            hx = hx - 1;
        }        
    }    
};


var createCircleTable = function() {

    //stroke(40, 186, 7);
    //translate(200,200);
    
    var x1 = 0;
    var y1 = radius;
    
    for (var deg = 0; deg < 90; deg += sLong)
    {
        var x2 = sin(deg + sLong) * radius;
        var y2 = cos(deg + sLong) * radius;

        fillCircleTable(x1, y1, x2,y2);

        x1 = x2;
        y1 = y2;
    }
    
    // HACK
    circleTab[0]=90;
};


var drawGrid = function(size) {

    translate (30,30);
    var cx = bx+0;
    var cy = by-34; 
    
    stroke(0, 133, 47);

    var t;
    var cpos = new PVector(cx,cy);

    // Vertical lines
    
    for (var vlines = 0, xpos = 0; vlines < 8; vlines++) {

        if (xpos < (cpos.x-radius) || xpos > (cpos.x + radius)) {
            line(xpos, 0, xpos, size);
        }
        else
        {
            var bspacex = cx - xpos;
            var index = floor(abs(bspacex));
            var circleY = circleTab[index];
            line (xpos, 0, xpos, max(cy-circleY,0));
            line (xpos, min(circleY+cy,size), xpos, size);
        }
        xpos+=(size/7);
    }
    
    //for (var vlines = 0, xpos = 0; vlines < 8; vlines++) {
        //line(xpos, size, xpos*1.05-8, size+8);
        //xpos+=(size/7);
    //} 

    // Horizontal lines    

    for (var hlines = 0, ypos = 0; hlines < 7; hlines++) {

        if (ypos < (cpos.y-radius) || ypos > (cpos.y + radius)) {
            line(0, ypos, size, ypos);
        }
        else {
            var bspacey = cy - ypos;
            var index = floor(abs(bspacey));
            var circleX = circleTab[index];
            line (0, ypos, max(cx-circleX-30,0),ypos);
            line (min(circleX+cx,size), ypos, size,ypos);
        }

        ypos+=(size/6);

    }
    
    //ypos-=(size/6)-2;
    //line(-1, ypos, size+2, ypos);
    //ypos+=4;
    //line(-5, ypos, size+6, ypos);


};

var drawCircumference = function() {

    //stroke(255, 60, 0);

    var bx = 0;
    var by = radius;

    var deg = 0;

    for (deg = 0; deg < (180-sLong); deg += sLong)
    {
        var tx = Math.round(sin(deg + sLong) * radius);
        var ty = Math.round(cos(deg + sLong) * radius);

        cline(tx, ty, bx, by);
        cline(-tx, ty, -bx, by);
        bx = tx;
        by = ty;

    }
    {
        var tx = Math.round(sin(deg + sLong) * radius);
        var ty = Math.round(cos(deg + sLong) * radius);
        cline(tx, ty, bx, by);
        cline(-tx, ty, -bx, by);
    }        

};

var drawLatitude = function() {

    //stroke(255, 60, 0);

    var deg = 0;
    var index = 1;

    for (deg = sLong; deg < (180.0); deg += sLong)
    {
        var x1 = Math.round(sin(deg) * radius);
        var y1 = Math.round(cos(deg) * radius);

        var x2 = Math.round(sin(deg + index * -45 ) * radius);
        var y2 = Math.round(cos(deg + index * -45) * radius);

        var rx1 = x1 * cosra - y1 * sinra;
        var ry1 = x1 * sinra + y1 * cosra;

        var rx2 = x2 * cosra - y2 * sinra;
        var ry2 = x2 * sinra + y2 * cosra;

        cline(rx1,ry1, rx2,ry2);

        index++;
    }

};


var drawLongitude = function( ani ) {

    //stroke(250, 0, 233);
    
    //rotate(-22.5);
    
    var down = true;

    for ( var latDeg = ani + (sLat /2); latDeg < 180- (sLat/2); latDeg += sLat )
    //for ( var latDeg = ani ; latDeg < 180; latDeg += sLat )
    {

        if (down)
        {
            for (var deg = 0; deg < 180; deg += sLong)
            {
                var x1 = cos(latDeg) * (radius * sin(deg));
                var y1 = cos(deg) * radius;
    
                var x2 = cos(latDeg) * (radius * sin(deg + sLong));
                var y2 = cos(deg+sLat) * radius;
                
                var rx1 = x1 * cosra - y1 * sinra;
                var ry1 = x1 * sinra + y1 * cosra;
    
                var rx2 = x2 * cosra - y2 * sinra;
                var ry2 = x2 * sinra + y2 * cosra;
    
                cline(Math.round(rx1), Math.round(ry1), Math.round(rx2), Math.round(ry2));
                //cline(Math.round(x1), Math.round(y1), Math.round(x2), Math.round(y2));
            }
        }
        else
        {
            for (var deg = 180; deg > 0; deg -= sLong)
            {
                var x1 = cos(latDeg) * (radius * sin(deg));
                var y1 = cos(deg) * radius;
    
                var x2 = cos(latDeg) * (radius * sin(deg + sLong));
                var y2 = cos(deg+sLat) * radius;
                
                var rx1 = x1 * cosra - y1 * sinra;
                var ry1 = x1 * sinra + y1 * cosra;
    
                var rx2 = x2 * cosra - y2 * sinra;
                var ry2 = x2 * sinra + y2 * cosra;
    
                cline(Math.round(rx1), Math.round(ry1), Math.round(rx2), Math.round(ry2));
            }
        }
        
        down = !down;
    }
};

var imagePrep = function(x, y) {
  
    var r = radius*2;
    pic.beginDraw();
    pic.background(0, 0, 0);
    
    pic.fill(255, 255, 255);
    pic.ellipse (x,y,r,r);
    
    pic.endDraw();
    
};

//debug("---------NEW RUN----------");

//background(0, 0, 0);
createCircleTable();
//println(circleTab);
//debug(circleTab[0]);
//debug(circleTab[90]);
//for (var xx = 0; xx < 90; xx++) {
//    debug (circleTab[xx]);
//}

var draw1 = function() {
};

var draw = function() {
    
    lpf = 0;
    tlen = 0;
    
    background(0, 0, 0);
    
    resetMatrix();

    //imagePrep(bx,by);
    //image(pic, 0, 0);
    //pic.loadPixels();
    //zbuffer = pic.imageData.data;

    strokeWeight(1);    
    drawGrid(342);

    resetMatrix();

    //point(bx,by);
    //stroke(235, 17, 68);
    //ellipse(200,200,110,110);
  
    stroke(29, 133, 66);
    strokeWeight(2);

    translate(bx,by);
    //rotate(22.5);

    drawCircumference();
    var ani = Math.trunc(bx*1.5%sLat);
    //println(ani);
    drawLongitude(ani);
    drawLatitude();

    bx = bx + xs;
    if (bx < 90){
        bx = bx - xs;
        xs = -xs;
    }
    if (bx > 310){
        bx =bx - xs;
        xs = -xs;
    }
    
    by = by + n;
    n = n + dn;
    if (by > 310 ){
        n = n - dn;
        n = - n;
    }
    
    //println(tlen);
    //println(lpf);
    
    
    
};

draw1();

